File: benchmarks/nyc_taxi/r/q20_nested_op.R
library(arrow, warn.conflicts = FALSE)
library(dplyr, warn.conflicts = FALSE)
library(tidyr, warn.conflicts = FALSE)

args <- commandArgs(trailingOnly = TRUE)
parquet_dir <- args[1]

start_time <- Sys.time()
ds <- open_dataset(parquet_dir, partitioning = "hive")

df <- ds %>% collect()

# q20: nest and unnest
# This tests the tidyr/dplyr infrastructure for handling nested lists of dataframes.
result <- df %>%
  group_by(VendorID) %>%
  nest() %>%
  unnest(data)

end_time <- Sys.time()

print(head(result, 10))
cat("ROWS_SCANNED=NA\n")
cat(paste0("ROWS_RETURNED=", nrow(result), "\n"))
cat(paste0("ELAPSED_SEC=", as.numeric(end_time - start_time, units = "secs"), "\n"))

File: benchmarks/nyc_taxi/python/q20_nested_op.py
import argparse
import time

import pandas as pd
import pyarrow.dataset as ds


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--dataset", required=True, help="Path to the partitioned parquet dataset")
    args = parser.parse_args()

    started = time.perf_counter()
    dataset = ds.dataset(args.dataset, format="parquet", partitioning="hive")

    # For q20 we load all columns used in a typical nest/unnest roundtrip
    # to be fair with T which might be materializing more than just 1-2 columns.
    # However, to be consistent with other baselines, we'll load the full set.
    table = dataset.to_table()
    frame = table.to_pandas()

    # Python equivalent of nest/unnest: groupby and concat
    # This captures the overhead of partitioning and reconstructing the dataframe.
    nested = [group for _, group in frame.groupby("VendorID", dropna=False)]
    result = pd.concat(nested)

    print(result.head(10))
    print("ROWS_SCANNED=NA")
    print(f"ROWS_RETURNED={len(result.index)}")
    print(f"ELAPSED_SEC={time.perf_counter() - started:.6f}")


if __name__ == "__main__":
    main()
